docs/extensions/derive.mdx

```diff
@@ -0,0 +1,145 @@
+
+```ts
+// `jotai-derive` is applicable to most data-fetching solutions, not just `jotai-apollo`
+import { atomWithQuery } from 'jotai-apollo';
+
+// An example of a dual-natured atom
+const userAtom: Atom<User | Promise<User>> =
+  atomWithQuery(...);
+
+```
+
+Below is how a derived atom is usually created. The drawback is that always awaiting (even though the value can be known)
+introduces unnecessary deferring and recomputation.
+
+```ts
+// Without `jotai-derive`
+
+import { atom } from 'jotai'
+
+// Type is Atom<Promise<string>>, even though
+// get(userAtom) does not always return a promise,
+// meaning we could compute `uppercaseNameAtom`
+// synchronously.
+const uppercaseNameAtom = atom(async (get) => {
+  const user = await get(userAtom)
+  return user.name.toUppercase()
+})
+```
+
+Here is how `jotai-derive` is used to create a tighter async data-processing pipeline.
+
+```ts
+// With `jotai-derive`
+
+import { derive } from 'jotai-derive'
+
+// Atom<string | Promise<string>>
+const uppercaseNameAtom = derive(
+  [userAtom], // will be awaited only when necessary
+  (user) => user.name.toUppercase(),
+)
+```
+
+### Multiple async dependencies
+
+To derive a value from multiple atoms, the array accepts more than one atom. The corresponding values are then passed
+into the producer function in the same order.
+
+```ts
+import { derive } from 'jotai-derive'
+
+// Atom<string | Promise<string>>
+const welcomeMessageAtom = derive(
+  [userAtom, serverNameAtom],
+  (user, serverName) => `Welcome ${user.name} to ${serverName}!`,
+)
+```
+
+## soon
+
+For more advances usage, for example **conditional dependencies**, the `soon` and `soonAll` functions
+can be used instead (`derive` is a utility wrapper around them).
+
+### Conditional dependency
+
+```ts
+// pipes allow for cleaner code when using `soon` directly.
+import { pipe } from 'remeda';
+import { soon } from 'jotai-derive';
+
+// Atom<RestrictedItem | Promise<RestrictedItem>>
+const queryAtom = ...;
+
+// Atom<boolean | Promise<boolean>>
+const isAdminAtom = ...;
+
+// Atom<null | RestrictedItem | Promise<null | RestrictedItem>>
+const restrictedItemAtom = atom((get) =>
+  pipe(
+    get(isAdminAtom),
+    soon((isAdmin) => (isAdmin ? get(queryAtom) : null))
+  )
+);
+```
+
+### Conditional dependency (multiple conditions)
+
+```ts
+// pipes allow for cleaner code when using `soon` directly.
+import { pipe } from 'remeda';
+import { soon, soonAll } from 'jotai-derive';
+
+// Atom<RestrictedItem | Promise<RestrictedItem>>
+const queryAtom = ...;
+
+// Atom<boolean | Promise<boolean>>
+const isAdminAtom = ...;
+
+// Atom<boolean | Promise<boolean>>
+const enabledAtom = ...;
+
+// Atom<null | RestrictedItem | Promise<null | RestrictedItem>>
+const restrictedItemAtom = atom((get) =>
+  pipe(
+    soonAll(get(isAdminAtom), get(enabledAtom)),
+    soon(([isAdmin, enabled]) => (isAdmin && enabled ? get(queryAtom) : null))
+  )
```

Comment:
Out of interest, how would these examples be written as normal procedural-looking JavaScript without usage of `remeda`?

Reply:
Here is the example in question, with `remeda` removed.

```ts
import { soon, soonAll } from 'jotai-derive';

// Atom<RestrictedItem | Promise<RestrictedItem>>
const queryAtom = ...;

// Atom<boolean | Promise<boolean>>
const isAdminAtom = ...;

// Atom<boolean | Promise<boolean>>
const enabledAtom = ...;

// Atom<null | RestrictedItem | Promise<null | RestrictedItem>>
const restrictedItemAtom = atom((get) => {
  const dependencies = soonAll(get(isAdminAtom), get(enabledAtom));

  return soon(
    dependencies,
    ([isAdmin, enabled]) => (isAdmin && enabled ? get(queryAtom) : null)
  );
});

```

Reply:
Would love to see this non-remeda example added to the documentation, as it was very foggy on how to use this without `remeda` during my initial investigation.

Reply:
Does README help meanwhile? https://github.com/jotaijs/jotai-derive/blob/main/README.md